chocolatey/choco
# (#3709) Add support for ordering strategies using a new `--order-by` argument on the search command

## Description Of Changes

This change adds a new command line option "order-by", allowing the user to specify a few sorting strategies. By default the results will be sorted by id.

## Motivation and Context

To allow uses to decide how they want the results being sorted when we can.

## Testing

These tests assume you only have the Chocolatey Community Repository enabled as a source.

1. Do a search using `choco search chrome --order-by='id'`
2. Verify the resulting set is ordered by the Identifier with the first entry being 1password-chrome
3. Do the same search using `--order-by='title'`
4. Verify the first entry is `choco-protocol-support` (It is prefixed with `(unofficial)` in the title, and is expected to be the first result in this case.
5. Search again using `--order-by='popularity'`
6. Verify the package with most downloads is returned first (at this time it is `GoogleChrome`).
7. Search again using `--order-by='unsorted'`
8. Go to https://community.chocolatey.org/packages` and enter `chrome` in the search box and hit enter.
9. Verify the order of the search on Chocolatey Community Repository and the results from Chocolatey CLI is in the same/similar order.
10. Search again using `--order-by='LastPublished'`.
11. Verify a yellow warning is outputted with the text `Order By 'LastPublished' has been specified. This order is done on client side and may lead to inconsistently ordered results.`
12. Go to the website search previously made, and change the dropdown for the order to `Recent`.
13. Verify the order of packages is similar to the output from the Chocolatey CLI output (they may not be exactly the same due to client sort).
14. Search again using `--order-by='other'`.
15. Verify an error is outputted with the message `Error: The order-by clause 'other' is not recognized. Use one of the supported clauses: 'Id', 'LastPublished', 'Popularity', 'Title', 'Unsorted'.`.

### Operating Systems Testing

- Windows 11

## Change Types Made


## Change Checklist


## Related Issue
Fixes #3709

## Additional Information

Note that some results may appear out-of-order, this is caused by the NuGet server returning them in the incorrect order, see https://github.com/chocolatey/NuGet.Server-chocolatey/issues/2
